Add webpushd IPC to query for all domains with active Web Push subscriptions

bnham opened the pull request:
#### bc1e128089c0d6569702b2232aef9100f82fb78a
<pre>
Add webpushd IPC to query for all domains with active Web Push subscriptions
<a href="https://bugs.webkit.org/show_bug.cgi?id=243347">https://bugs.webkit.org/show_bug.cgi?id=243347</a>
rdar://93193992

Reviewed by Alex Christensen.

Some parts of WebKit (e.g. ITP) need the ability to query for all domains with active Web Push
subscriptions. We should add an IPC to webpushd to facilitate this.

* Source/WebCore/Modules/push-api/PushDatabase.cpp:
(WebCore::PushDatabase::getOriginsWithPushSubscriptions):
* Source/WebCore/Modules/push-api/PushDatabase.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::getOriginsWithPushSubscriptions):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::getOriginsWithPushSubscriptions):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
(WebKit::WebPushD::messageTypeSendsReply):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _getOriginsWithPushSubscriptions:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getOriginsWithPushSubscriptions):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/webpushd/PushService.h:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::getOriginsWithPushSubscriptions):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::MessageInfo::getOriginsWithPushSubscriptions::encodeReply):
(WebPushD::handleWebPushDMessageWithReply):
(WebPushD::Daemon::decodeAndHandleMessage):
(WebPushD::Daemon::getOriginsWithPushSubscriptions):
* Tools/TestWebKitAPI/Tests/WebCore/PushDatabase.cpp:
(TestWebKitAPI::PushDatabaseTest::getOriginsWithPushSubscriptions):
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: <a href="https://commits.webkit.org/252970@main">https://commits.webkit.org/252970@main</a>
</pre>

achristensen07 commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -558,6 +573,18 @@ void handleWebPushDMessage(ClientConnection* connection, Span<const uint8_t> enc
+        if (!m_pushService)
+            return callback({ });
+
+        m_pushService->getOriginsWithPushSubscriptions(bundleIdentifier, [callback = WTFMove(callback)](auto&& origins) mutable {
```
I don't see why you wrapped callback in a lambda.

bnham replied:
It's because one of the lambdas is of type `CompletionHandler<void(const Vector<String>&)>` and the other is of type `CompletionHandler<void(Vector<String>&&)>`.

bnham replied:
Ok, I fixed the `REPLY` macro to work with movable types, so this is cleaned up now.

bnham committed:
Add webpushd IPC to query for all domains with active Web Push subscriptions
https://bugs.webkit.org/show_bug.cgi?id=243347
rdar://93193992

Reviewed by Alex Christensen.

Some parts of WebKit (e.g. ITP) need the ability to query for all domains with active Web Push
subscriptions. We should add an IPC to webpushd to facilitate this.

* Source/WebCore/Modules/push-api/PushDatabase.cpp:
(WebCore::PushDatabase::getOriginsWithPushSubscriptions):
* Source/WebCore/Modules/push-api/PushDatabase.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::getOriginsWithPushSubscriptions):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::getOriginsWithPushSubscriptions):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
(WebKit::WebPushD::messageTypeSendsReply):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _getOriginsWithPushSubscriptions:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getOriginsWithPushSubscriptions):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/webpushd/PushService.h:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::getOriginsWithPushSubscriptions):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::MessageInfo::getOriginsWithPushSubscriptions::encodeReply):
(WebPushD::handleWebPushDMessageWithReply):
(WebPushD::Daemon::decodeAndHandleMessage):
(WebPushD::Daemon::getOriginsWithPushSubscriptions):
* Tools/TestWebKitAPI/Tests/WebCore/PushDatabase.cpp:
(TestWebKitAPI::PushDatabaseTest::getOriginsWithPushSubscriptions):
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: https://commits.webkit.org/252970@main
